google-cloud-storage/src/main/java/com/google/cloud/storage/BaseStorageReadChannel.java

```diff
@@ -0,0 +1,149 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import static com.google.cloud.storage.ByteSizeConstants._2MiB;
+import static java.util.Objects.requireNonNull;
+
+import com.google.cloud.storage.BufferedReadableByteChannelSession.BufferedReadableByteChannel;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+abstract class BaseStorageReadChannel<T> implements StorageReadChannel {
+
+  private ByteRangeSpec byteRangeSpec;
+  private int chunkSize = _2MiB;
+  private BufferHandle bufferHandle;
+  private LazyReadChannel<T> lazyReadChannel;
+
+  @Nullable private T resolvedObject;
+
+  protected BaseStorageReadChannel() {
+    this.byteRangeSpec = ByteRangeSpec.nullRange();
+  }
+
+  @Override
+  public final synchronized void setChunkSize(int chunkSize) {
+    StorageException.wrapIOException(() -> maybeResetChannel(true));
+    this.chunkSize = chunkSize;
+  }
+
+  @Override
+  public final synchronized boolean isOpen() {
+    if (lazyReadChannel == null) {
+      return true;
+    } else {
+      LazyReadChannel<T> tmp = internalGetLazyChannel();
+      return tmp.isOpen();
+    }
+  }
+
+  @Override
+  public final synchronized void close() {
+    if (internalGetLazyChannel().isOpen()) {
+      StorageException.wrapIOException(internalGetLazyChannel().getChannel()::close);
+    }
+  }
+
+  @Override
+  public final synchronized StorageReadChannel setByteRangeSpec(ByteRangeSpec byteRangeSpec) {
+    requireNonNull(byteRangeSpec, "byteRangeSpec must be non null");
+    StorageException.wrapIOException(() -> maybeResetChannel(false));
+    this.byteRangeSpec = byteRangeSpec;
+    return this;
+  }
+
+  @Override
+  public final ByteRangeSpec getByteRangeSpec() {
+    return byteRangeSpec;
+  }
+
+  @Override
+  public final synchronized int read(ByteBuffer dst) throws IOException {
+    long diff = byteRangeSpec.length();
+    if (diff <= 0) {
+      close();
+      return -1;
+    }
+    try {
+      int read = internalGetLazyChannel().getChannel().read(dst);
+      if (read != -1) {
+        byteRangeSpec = byteRangeSpec.withShiftBeginOffset(read);
+      } else {
+        close();
+      }
+      return read;
+    } catch (StorageException e) {
+      if (e.getCode() == 416) {
+        // HttpStorageRpc turns 416 into a null etag with an empty byte array, leading
+        // BlobReadChannel to believe it read 0 bytes, returning -1 and leaving the channel open.
+        // Emulate that same behavior here to preserve behavior compatibility, though this should
+        // be removed in the next major version.
+        return -1;
+      } else {
+        throw new IOException(e);
+      }
+    } catch (IOException e) {
+      throw e;
+    } catch (Exception e) {
```

Comment:
This is broad; is there a specific set of Exception types that are expected?

Reply:
For grpc, the scope is more narrow, but for JSON there isn't really a common base since things can be socket issues, ssl issues or any other number of exceptions. 

This catch is primarily here to guarantee we're always doing our best to present an actionable exception to whatever is catching.